doc/tutorials/debug.rst

```diff
@@ -0,0 +1,322 @@
+
+1. Start NPK output in SOS::
+
+      $ echo 1 > /sys/bus/intel_th/devices/0-msc0/active
+
+2. Wait a moment, stop NPK output, and get the log from memory::
+
+      $ echo 0 > /sys/bus/intel_th/devices/0-msc0/active
+      $ cat /dev/intel_th0/msc0 > npk_trace
+
+3. Use `PVT
+   <https://ubit-artifactory.intel.com/artifactory/list/dsdjf-pvt-repo/1909/Red/InstallPVT.All.1.0.9006.600.exe>`_
+   tool to load the NPK log:
+
+   .. figure:: images/debug_image26.png
+      :align: center
+
+      load NPK log
+
+   Then you will see the NPK log as follows:
+
+   .. figure:: images/debug_image27.png
+      :align: center
+
+      NPK log
+
+ACRN Trace
+**********
+
+ACRN trace is a tool running on the Service OS (SOS) to capture trace
+data. We can use the existing trace information to analyze, and we can
+add self-defined tracing to analyze code which we care about.
+
+Using Existing trace event id to analyze trace
+==============================================
+
+As an example, we can use the existing vm_exit trace to analyze the
+reason and times of each vm_exit after we have done some operations.
+
+1. Run the following SOS console command to collect
+   trace data::
+
+      # acrntrace -c
+
+2. Check current directory, and confirm the directory contains four
+   trace files::
+
+      # ls
+      0 1 2 3
+
+3. Use the command to get a summary of vmexit::
+
+      # acrnalyze.py -i /home/trace/acrntrace/20190219-001529/1 -o vmexit --vm_exit
+
+   .. figure:: images/debug_image28.png
+      :align: center
+
+      vmexit summary information
+
+Using Self-defined trace event id to analyze trace
+==================================================
+
+For some undefined trace event id, we can define it by ourselves as
+shown in the following example:
+
+1. Add the following new event id into
+   ``acrn-hypervisor/hypervisor/include/debug/trace.h``:
+
+   .. figure:: images/debug_image25.png
+      :align: center
+
+      trace event id
+
+2. Add the following format to
+   ``acrn-hypervisor/tools/acrntrace/scripts/formats``:
+
+   .. figure:: images/debug_image1.png
+      :align: center
+
+      acrntrace formatted information
+
+   .. note::
+
+      Formats:
+        0x00000005: event id for trace test
+
+        %(cpu)d: corresponding cpu index with 'decimal' format
+
+        %(event)016x: corresponding event id with 'hex' format
+
+        %(tsc)d: corresponding event time stamp with 'decimal' format
+
+        %(1)08x: corresponding first 'Long' data in TRACE_2L
+
+3. Add trace into function ``emulate_io`` in
+   ``acrn-hypervisor/hypervisor/arch/x86/guest/io_emul.c`` which we want to
+   trace for the calling times of function ``emulate_io``:
+
+   .. figure:: images/debug_image2.png
+      :align: center
+
+      inserted trace information
+
+4. After we have inserted the trace code addition, we need to rebuild
+   the acrn-hypervisor and acrn-kernel, and flash into the board::
+
+      # make
+      # make flash
```

Comment:
This does not work on the open-source tree, we should add a note to that effect if we want to keep these instructions in here.